cpp/src/arrow/array/builder_binary.h

```diff
@@ -303,6 +304,77 @@ class BaseBinaryBuilder
     return Status::OK();
   }
 
+  Status AppendBinaryWithLengths(std::string_view binary, const int32_t* value_lengths,
+                                 int64_t length) {
```

Comment:
Why not call this `AppendValuesWithLengths`?
Also, it should probably take the right offset type?
And you need to add a docstring...
```suggestion
  /// XXX docstring
  Status AppendValuesWithLengths(std::string_view binary, util::span<const offset_type> lengths) {
```